cnd/src/btsieve/ethereum/mod.rs

```diff
@@ -24,16 +20,134 @@ use std::collections::HashSet;
 type Hash = H256;
 type Block = crate::ethereum::Block<Transaction>;
 
-pub async fn matching_transaction<C>(
+pub const TRANSACTION_STATUS_OK: u32 = 1;
+
+pub async fn matching_create_contract_transaction<C>(
+    blockchain_connector: C,
+    start_of_swap: NaiveDateTime,
+    bytecode: Bytes,
+) -> anyhow::Result<Transaction>
+where
+    C: LatestBlock<Block = Option<Block>>
+        + BlockByHash<Block = Option<Block>, BlockHash = Hash>
+        + ReceiptByHash<Receipt = Option<TransactionReceipt>, TransactionHash = Hash>
+        + Clone,
+{
+    matching_transaction(blockchain_connector.clone(), start_of_swap, |transaction| {
+        Box::new({
+            let blockchain_connector = blockchain_connector.clone();
+            let bytecode = bytecode.clone();
+            async move {
+                // transaction.to address is None if, and only if, the transaction
+                // creates a contract.
+                if transaction.to.is_none() && transaction.input == bytecode {
+                    // We can get a transaction that matches even for failed attempts
+                    // at contract creation e.g., low gas.
+                    let receipt = fetch_receipt(blockchain_connector, transaction.hash).await?;
+                    if receipt.status == Some(TRANSACTION_STATUS_OK.into()) {
```

Comment:
Oh, nice one! I didn't think of that!

This got me thinking:
We should probably never return any transaction that failed right?

So the whole check should probably move into `matching_transaction` and run _after_ the matcher returned true. This will make this matcher non-async again :)
We probably then want to log a big warning that says:

"We found the transaction we were looking for but the execution failed!"

Reply:
This would also reduce some code duplication between the matchers.